### PR TITLE
Fixes liveness analysis issue in looplifting

### PR DESCRIPTION
**Pull request by stuartarchibald**
This is a part fix for looplifting failing to compute liveness
at block exit correctly.

### CODE REVIEW

File: numba/tests/test_looplifting.py [hunk 1]
```diff
@@ -378,12 +379,42 @@ def test(n):
                 res = i
             return res
 
+        # loop count = 0, loop won't lift or execute
+        self.assertEqual(test.py_func(-1), test(-1))
+
+        # loop count = 0, loop won't lift but will execute
```
Comment (sklam) [inline numba/tests/test_looplifting.py]: loop count = 1?

Comment (stuartarchibald) [reply numba/tests/test_looplifting.py]: It won't lift following this change.

**Comment by stuartarchibald** [comment]
Without the proposed patch, anything defined in the loop cannot leave, e.g. this fails:

```python
from numba import jit

@jit(forceobj=True)
def test(n):
    for i in range(n):
        res = i
    return res
test(1)
```
with 
```
  File "<path>/numba/numba/objmode.py", line 520, in loadvar
    assert name in self._live_vars, name
AssertionError: res
```
if the defines from the loop body are added to the lifted loop outputs then the above will run.
(as done here https://github.com/numba/numba/pull/3230/files#diff-256b76c54a3223e4b07f4148205ddae4R120)

However this sort of thing then causes issues:
```python
@jit(forceobj=True)
def test(n):
    res = 0
    for i in range(n):
        res = i
    return res

test(1) # WORKS
test(-1) # BROKEN - `NameError: name 'res' is not defined`
```
Dumping out some info for the loop above (with the defines added to outputs) gives:
```
0
    n = arg(0, name=n)                       ['n']
    $const0.1 = const(int, 0)                ['$const0.1']
    res = $const0.1                          ['$const0.1', 'res']
    del $const0.1                            []
    jump 4                                   []
--> Used:
--> []
--> Defined:
--> ['$const0.1', 'n', 'res']
--> Livemap:
--> set()
4
    jump 6                                   []
--> Used:
--> []
--> Defined:
--> []
--> Livemap:
--> {'n', 'res'}
6
    $6.1 = global(range: <class 'range'>)    ['$6.1']
    $6.3 = call $6.1(n, func=$6.1, args=[Var(n, foo4.py (5))], kws=(), vararg=None) ['$6.1', '$6.3', 'n']
    del n                                    []
    del $6.1                                 []
    $6.4 = getiter(value=$6.3)               ['$6.3', '$6.4']
    del $6.3                                 []
    $phi14.1 = $6.4                          ['$6.4', '$phi14.1']
    del $6.4                                 []
    jump 14                                  []
--> Used:
--> ['n']
--> Defined:
--> ['$6.1', '$6.3', '$6.4', '$phi14.1']
--> Livemap:
--> {'n', 'res'}
14
    $14.2 = iternext(value=$phi14.1)         ['$14.2', '$phi14.1']
    $14.3 = pair_first(value=$14.2)          ['$14.2', '$14.3']
    $14.4 = pair_second(value=$14.2)         ['$14.2', '$14.4']
    del $14.2                                []
    $phi16.1 = $14.3                         ['$14.3', '$phi16.1']
    $phi24.1 = $14.3                         ['$14.3', '$phi24.1']
    del $phi24.1                             []
    del $14.3                                []
    $phi24.2 = $phi14.1                      ['$phi14.1', '$phi24.2']
    del $phi24.2                             []
    branch $14.4, 16, 24                     ['$14.4']
--> Used:
--> ['$phi14.1']
--> Defined:
--> ['$14.2', '$14.3', '$14.4', '$phi16.1', '$phi24.1', '$phi24.2']
--> Livemap:
--> {'res', '$phi14.1'}
16
    del $14.4                                []
    i = $phi16.1                             ['$phi16.1', 'i']
    del $phi16.1                             []
    res = i                                  ['i', 'res']
    del i                                    []
    jump 14                                  []
--> Used:
--> ['$phi16.1']
--> Defined:
--> ['i', 'res']
--> Livemap:
--> {'$phi16.1', '$phi14.1'}
24
    del $phi16.1                             []
    del $phi14.1                             []
    del $14.4                                []
    jump 26                                  []
--> Used:
--> []
--> Defined:
--> []
--> Livemap:
--> {'res'}
26
    $26.2 = cast(value=res)                  ['$26.2', 'res']
    del res                                  []
    return $26.2                             ['$26.2']
--> Used:
--> ['res']
--> Defined:
--> ['$26.2']
--> Livemap:
--> {'res'}

Loop Info:
Inputs:
['n', 'res']
Outputs:
['res']
Used:
['$phi14.1', '$phi16.1', 'n']
Defined:
['$14.2', '$14.3', '$14.4', '$6.1', '$6.3', '$6.4', '$phi14.1', '$phi16.1', '$phi24.1', '$phi24.2', 'i', 'res']
```
`res` is considered an input, it is also defined, not used and is an output, this seems valid.

Once the looplifting is done the IR looks like:
```
label 0:
    n = arg(0, name=n)                       ['n']
    $const0.1 = const(int, 0)                ['$const0.1']
    res = $const0.1                          ['$const0.1', 'res']
    del res                                  []
    del $const0.1                            []
    jump 4                                   []
label 4:
    jump 6                                   []
label 6:
    $19 = const(LiftedLoop, LiftedLoop(<function test at 0x7f2994a4d268>)) ['$19']
    $20 = call $19(n, func=$19, args=[Var(n, foo4.py (5))], kws=(), vararg=None) ['$19', '$20', 'n']
    del n                                    []
    del $19                                  []
    res = static_getitem(value=$20, index=0, index_var=None) ['$20', 'res']
    del $20                                  []
    jump 26                                  []
label 26:
    $26.2 = cast(value=res)                  ['$26.2', 'res']
    del res                                  []
    return $26.2                             ['$26.2']
```
`res` is deleted in block 0, so if the lifted loop does not execute there's no definition of `res`.


This patch proposes rejection of any loop that has a redefinition of a variable that is in the lifted loop inputs as a part fix, it's essentially a workaround.

I think the actual problem is that in postproc for the mutated IR the variable `res` is not in the live map for block 4 or 6:
```
(Pdb) p sorted(vlt.livemap.items())
[(0, set()), (4, {'n'}), (6, {'n'}), (26, {'res'})]
```
but the usedefs seem ok, but inconsistent with the notion of the lifted loop having inputs `['n', 'res']` which implies use of `res` in the loop?
```
(Pdb) p sorted(vlt.usedefs.usemap.items())
[(0, set()), (4, {'n'}), (6, {'n'}), (26, {'res'})]
(Pdb) p sorted(vlt.usedefs.defmap.items())
[(0, {'res', '$const0.1', 'n'}), (4, set()), (6, {'res', '$19', '$20'}), (26, {'$26.2'})]
```

**Comment by sklam** [comment]
@stuartarchibald, I discovered that there was an error with the liveness analysis causing it to give incorrect liveness info (i.e. variables that are live before they are actually defined).  Once fixed, I can avoid the need for `cannot_redefine()`.  See

https://github.com/stuartarchibald/numba/compare/fix/liveness_analysis_issue...sklam:pr/alt/3230_2?expand=1

**Comment by stuartarchibald** [comment]
@sklam thanks for looking into this, seems like the right fix, feel free to push to this branch. Thanks.

**Comment by sklam** [comment]
(It's weird that the lately pushed commit goes before the comment about pushing those commits)

**Comment by stuartarchibald** [comment]
This should close (as I've manually checked all these cases):
* https://github.com/numba/numba/issues/3204
* https://github.com/numba/numba/issues/3029
* https://github.com/numba/numba/issues/3078
* https://github.com/numba/numba/issues/2715
* https://github.com/numba/numba/issues/2254

But doesn't fix:
* https://github.com/numba/numba/issues/3131

File: numba/tests/test_looplifting.py [hunk 2]
```diff
@@ -378,11 +384,43 @@ def test(n):
                 res = i
             return res
 
-        # loop count = 0
+        # loop count = 1, loop lift but loop body not execute
+        self.assertEqual(test.py_func(-1), test(-1))
+        self.assert_has_lifted(test, loopcount=1)
+        # loop count = 1, loop won't lift and will execute
```
Comment (stuartarchibald) [inline numba/tests/test_looplifting.py]: s/won't/will/

**Comment by stuartarchibald** [comment]
Thanks for fixing the liveness analysis @sklam, this seems to now work as expected in cases we knew were problematic. As well as having the unit tests I've also tested all those user bug report cases manually and all now pass. I wonder if we should at some point add some tests that just test the postproc on handcrafted IR?

I think there's now one tiny typo https://github.com/numba/numba/pull/3230#pullrequestreview-149361861 where a test wasn't working that now is, on fixing that I think this can be merged.

**Pull request merged**